clippy_lints/src/append_instead_of_extend.rs

```diff
@@ -0,0 +1,69 @@
+use clippy_utils::diagnostics::span_lint_and_help;
+use clippy_utils::ty::{is_type_diagnostic_item, is_type_lang_item};
+use rustc_hir::{Expr, ExprKind, LangItem};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::symbol::sym;
+
+declare_clippy_lint! {
+    /// **What it does:** finds occurrences of `vec.extend(other_vec.drain(..))
+    ///
+    /// **Why is this bad?** Slightly more succinct code and faster code
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// // example code where clippy issues a warning
+    /// vec![].extend(vec![1, 2, 3].drain(..))
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// // example code which does not raise clippy warning
+    /// vec![].append(&mut vec![])
+    /// ```
+    pub APPEND_INSTEAD_OF_EXTEND,
+    perf,
+    "default lint description"
+}
+
+declare_lint_pass!(AppendInsteadOfExtend => [APPEND_INSTEAD_OF_EXTEND]);
+
+impl<'tcx> LateLintPass<'tcx> for AppendInsteadOfExtend {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        //validate target object
+        if let ExprKind::MethodCall(method_name, _, exprs, _) = expr.kind {
+            //check source object
+            if let ExprKind::MethodCall(src_method, _, src_args, _) = &exprs[1].kind {
+                //check drain range
+                let ty = cx.typeck_results().expr_ty(&src_args[1]);
+                let ty = ty.peel_refs();
+                let range_check = is_type_lang_item(cx, ty, LangItem::RangeFull);
+
+                //check types
+                let ty = cx.typeck_results().expr_ty(&exprs[0]).peel_refs();
+                let vector_1 = is_type_diagnostic_item(cx, ty, sym::vec_type);
+
+                let src_ty = cx.typeck_results().expr_ty(&src_args[0]).peel_refs();
+                let vector_2 = is_type_diagnostic_item(cx, src_ty, sym::vec_type);
+
+                if method_name.ident.as_str() == "extend"
+                    && src_method.ident.as_str() == "drain"
+                    && range_check
+                    && vector_1
+                    && vector_2
+                {
+                    span_lint_and_help(
+                        cx,
+                        APPEND_INSTEAD_OF_EXTEND,
+                        expr.span,
+                        "use of `vec.extend(other_vec.drain(..))`",
+                        None,
+                        "consider using `vec![].append(&mut vec![]);` instead",
```

Comment:
Here you can go one step further by not just using `vec![]` but the actual name of the used objects via rusts `snippet` functions. [Here is an example in practice -->](https://github.com/rust-lang/rust-clippy/blob/21083875d211c29fcfa4a21fcd66d4601d2b618b/clippy_lints/src/methods/string_extend_chars.rs#L36-L41) The applicability is a nice bonus but definitely not required, so you can ignore it if you like :)

Reply:
Added this. Thanks for the suggestion